src/go/rpk/pkg/cli/topic/analyze.go

```diff
@@ -0,0 +1,743 @@
+	}
+
+	// Creating and closing the kafka client in this method is important as the goroutines it
+	// creates concurrently access `a.tpInfo`.
+	cl, err := clCreationFn(opts)
+	if err != nil {
+		return err
+	}
+	defer cl.Close()
+
+	for _, tp := range c.cfg.TopicPartitions {
+		_, exists := c.cfg.ConsumeRanges.StartOffsets.Lookup(tp.Topic, tp.Partition)
+		if exists {
+			c.tpsRemaining[tp] = struct{}{}
+		}
+	}
+
+	for {
+		fetches := cl.PollFetches(ctx)
+		if err := fetches.Err(); err != nil {
+			return err
+		}
+
+		// Note that the fetched records are not being read here. Instead two client hooks are registered
+		// "HookFetchBatchRead" and "HookFetchRecordBuffered".
+		//
+		// The "HookFetchBatchRead" is the only way to get the number of batches read and the compressed
+		// batch size. This hook, however, is called for every internally read batch regardless of whether
+		// its records end up being returned from "PollFetches" or not.
+		//
+		// Hence "HookFetchRecordBuffered" is used to ensure that all records have been read for every batch
+		// "HookFetchBatchRead" is called for.
+
+		completedTPs := c.checkForCompletedTPs()
+		cl.PauseFetchPartitions(completedTPs)
+
+		if c.isDone() {
+			return nil
+		}
+	}
+}
+
+type collectResult struct {
+	TopicPartitionInfo tpInfoMap `json:"topicPartitionInfo" yaml:"topic_partition_info"`
+	TimeRange          timeRange `json:"timeRange" yaml:"time_range"`
+}
+
+func (c *collector) results() collectResult {
+	return collectResult{c.tpInfo, c.cfg.TimeRange}
+}
+
+type percentileValue[V any] struct {
+	percentile float64
+	value      V
+}
+
+type topicSummary struct {
+	Topic            string
+	Partitions       int
+	BatchesPerSecond []percentileValue[float64]
+	BatchSizes       []percentileValue[int] // bytes
+}
+
+func (cr *collectResult) summarizeTopics(percentiles []float64) ([]topicSummary, error) {
+	totalDuration := cr.TimeRange.End.Sub(cr.TimeRange.Start)
+	tps := map[string]struct {
+		Partitions []int
+		BatchSizes []int
+		BatchRates []float64
+	}{}
+	for tp, info := range cr.TopicPartitionInfo {
+		e := tps[tp.Topic]
+		e.Partitions = append(e.Partitions, int(tp.Partition))
+
+		if info.empty() {
+			tps[tp.Topic] = e
+			continue
+		}
+
+		batchesPerS := float64(info.BatchesRead) / totalDuration.Seconds()
+		e.BatchRates = append(e.BatchRates, batchesPerS)
+
+		if !slices.IsSorted(info.BatchSizes) {
+			slices.Sort(info.BatchSizes)
+		}
+
+		p50, err := utils.Percentile(info.BatchSizes, 50.0)
+		if err != nil {
+			return nil, err
+		}
+		// Note that we insert the p50 batch size for each partition here.
+		// Another option could be inserting every sampled batch size then
+		// take percentiles of that. However, since each partition may have
+		// a different number of sampled batches care would need to be taken
+		// to insert the same number of samples per partition to avoid bias.
+		e.BatchSizes = append(e.BatchSizes, p50)
+
+		tps[tp.Topic] = e
+	}
+
+	tss := []topicSummary{}
```

Comment:
```suggestion
	var tss []topicSummary
```

Reply:
Good point, I assume this avoids the initial heap allocation for an empty slice?